client/dfclient/src/fov.cpp

```diff
@@ -0,0 +1,383 @@
+#include <cassert>
+#include <cmath>
+#include <set>
+#include <queue>
+#include <functional>
+
+#include "fov.hpp"
+
+namespace fov {
+
+static const bool doDeoverlap = true;
+
+static std::vector<segment> deoverlap(ncine::Vector2f origin, std::vector<segment> segments);
+
+inline static double cross(ncine::Vector2f a, ncine::Vector2f b) {
+    return a.x * b.y - a.y * b.x;
+}
+
+inline static double dot(ncine::Vector2f a, ncine::Vector2f b) {
+    return a.x * b.x + a.y * b.y;
+}
+
+static bool isOrientedClockwise(const chain& c) {
+    if (c.empty()) {
+        return true;
+    }
+
+    // Calculate the signed area of the polygon bounded by the chain.
+    // Assumption: there are no self-intersections in the chain.
+    float area = 0;
+    ncine::Vector2f prev = c.back();
+    for (ncine::Vector2f curr : c) {
+        area += cross(prev, curr);
+        prev = curr;
+    }
+
+    // In this convention, the polygonal chain is CW iff the signed area >= 0.f.
+    return area >= 0.f;
+}
+
+// Adapted from Tim Voght's implementation of Paul Bourke's circle intersection
+// algorithm:
+// http://paulbourke.net/geometry/circlesphere/tvoght.c
+// Returns the intersection points of the circles. Always returns at most
+// two points.
+// `out` must point to an at-least-two-element array.
+static size_t circleIntersection(const circle& a, const circle& b, ncine::Vector2f* out) {
```

Comment:
This case shouldn't happen. Look at the place where this function is used, and at the circles that are intersected with each other.

We are given a circle with its center at point `C` and with radius `r`, and the player position `O`. Then, we construct another circle with center `C'` being the center of the segment `|CO|`, and radius `r' = |CO| / 2`. Unless `O` is inside the given circle (which shouldn't happen now because the player always collides with the circle), those two circles will definitely overlap.

An illustration:

![illustration](https://user-images.githubusercontent.com/4950936/103585432-537a5180-4ee3-11eb-9bc7-7f4d77182245.png)

Even if the case you described happens, the function should just return two points which are the same. The outline will contain a degenerate segment (segment with coinciding ends). I doubt it will cause glitches, because the shadow cast by it shouldn't be visible at all (I may be wrong though, didn't test this).

Reply:
Ok. The image convinces me